### PR TITLE
move deprecated keys check sooner in checkpoint loading

### LINKED ISSUE #9403: Check for deprecated keys sooner in checkpoint loading
## Proposed refactoring or deprecation
Fail faster if the model checkpoint contains deprecated keys

### Motivation

We currently wait all the way until [`restore_callbacks`](https://github.com/PyTorchLightning/pytorch-lightning/blob/a079d7fccc0a9be25b40296f2a348c4b4f40c8cf/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L167-L179) to validate that a checkpoint dictionary doesn't contain deprecated keys. However, this happens after the initial checkpoint load, datamodule, and lightning module weights are resumed. 

It'd be better to fail faster immediately after the checkpoint load if the checkpoint contains deprecated keys.

### Pitch

Add a utility function in the checkpoint connector to handle loading & validation
```
def _load_and_validate(checkpoint_path: _PATH) -> Dict:
    checkpoint = self.trainer.training_type_plugin.load_checkpoint(checkpoint_path)
    if any(key in checkpoint for key in DEPRECATED_CHECKPOINT_KEYS):
            raise ValueError(
                "The checkpoint you're attempting to load follows an"
                " outdated schema. You can upgrade to the current schema by running"
                f" `python -m pytorch_lightning.utilities.upgrade_checkpoint --file {checkpoint_path}`"
            )
```    
and remove this check from `restore_callbacks`

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

Fixes #9403.

Move deprecated keys check from sooner from `restore_callbacks` to `resume_start`.
- see [comment](https://github.com/PyTorchLightning/pytorch-lightning/issues/9403#issuecomment-916254489) / #9396, error may be replaced with `upgrade(checkpoint)` in the future

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -62,7 +62,18 @@ def resume_start(self) -> None:
             return
 
         rank_zero_info(f"Restoring states from the checkpoint path at {checkpoint_path}")
-        self._loaded_checkpoint = self.trainer.training_type_plugin.load_checkpoint(checkpoint_path)
+        self._loaded_checkpoint = self._load_and_validate_checkpoint(checkpoint_path)
+
+    def _load_and_validate_checkpoint(self, checkpoint_path: _PATH) -> Optional[Dict[str, Any]]:
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:67]: let's use this here too: https://github.com/PyTorchLightning/pytorch-lightning/blob/ec828b826717cd3b5beabcb6d0cacf41b2320a98/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L147-L149

**Review by ananthsub — COMMENTED** [review]
This looks reasonable to me as an easy mitigation to fail faster while preserving the existing behavior. I do not think it's at odds with @awaelchli 's proposal around upgrading the checkpoint at load time to avoid this error raising altogether and delete this code

**Review by ananthsub — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
